### PR TITLE
fix: handle edge case and test for fibonacci_bottom_up

### PR DESCRIPTION
**Pull request by tapasrnk**
#### Description of Change
- Added test cases for Fibonacci_bottom_up
- Formatted the code to follow the contribution guideline.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: handle edge case and test for fibonacci_bottom_up

### CODE REVIEW

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 1]
```diff
@@ -1,6 +1,34 @@
...
+ * Space reduction :
+ * for i the fibonacci store only i-1 th and i-2 th state.
+ */
+#include <cassert>  /// for assert
 #include <iostream>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
 using namespace std;
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:27]: You might need to update all `std::` instances.
```suggestion
```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 2]
```diff
@@ -11,11 +39,13 @@ int fib(int n) {
     }
     return res[1];
 }
+static void test() {
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:42]:
```suggestion

/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 3]
```diff
@@ -11,11 +39,13 @@ int fib(int n) {
...
+    assert(fib(-2) == -1);
+    assert(fib(0) == 0);
+    assert(fib(1) == 1);
+    assert(fib(17) == 1597);
+}
 int main(int argc, char const *argv[]) {
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:48]:
```suggestion

/**
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
 * @returns 0 on exit
 */
int main(int argc, char const *argv[]) {
```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 4]
```diff
@@ -11,11 +39,13 @@ int fib(int n) {
...
-    int n;
-    cout << "Enter n: ";
-    cin >> n;
-    cout << "Fibonacci number is ";
-    cout << fib(n) << endl;
+    test();
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:49]:
```suggestion
    test();  // run self-test implementations
```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 5]
```diff
@@ -1,6 +1,34 @@
...
+ * for i the fibonacci store only i-1 th and i-2 th state.
+ */
+#include <cassert>  /// for assert
 #include <iostream>
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
 using namespace std;
 int fib(int n) {
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:28]: Would be awesome if you could document this function according to our guidelines. If you need any help, let us know. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Good work! 👍 Let's take the opportunity to fix some of the warnings and improve the code.

**Comment by tapasrnk** [comment]
I have resolved all the requested changes and pushed some commits.

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 6]
```diff
@@ -1,6 +1,36 @@
...
+ * ### Algorithm
+ * The idea of the algorithm is to find the ith fibonacci number.
+ * DP state : F(i) = ith fibonacci number
+ * DP state transition :
+ *      F(i) = F(i-1) + F(i-2)
+ * DP base case :
+ *      F(0) = 0
+ *      F(1) = 1
+ * Space reduction :
+ * for i the fibonacci store only i-1 th and i-2 th state.
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:19]:
```suggestion
 * for i the fibonacci store only i-1 th and i-2 th state.

 * @author [Santo Cariotti](https://github.com/dcariotti)
 * @author [tapasrnk](https://github.com/tapasrnk)
```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 7]
```diff
@@ -1,6 +1,36 @@
...
+ */
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:26]:
```suggestion
 */
namespace dynamic_programming {
```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 8]
```diff
@@ -11,11 +41,23 @@ int fib(int n) {
     }
     return res[1];
 }
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:43]:
```suggestion
}
}  // namespace dynamic_programming

```

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 9]
```diff
@@ -11,11 +41,23 @@ int fib(int n) {
...
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    assert(fib(-2) == -1);
+    assert(fib(0) == 0);
+    assert(fib(1) == 1);
+    assert(fib(17) == 1597);
+}
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:53]:
```suggestion
    assert(dynamic_programming::fib(-2) == -1);
    assert(dynamic_programming::fib(0) == 0);
    assert(dynamic_programming::fib(1) == 1);
    assert(dynamic_programming::fib(17) == 1597);

    std::cout << "All tests have successfully passed!\n";
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 🙂

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 10]
```diff
@@ -1,6 +1,36 @@
...
+ * @brief Dynamic Programming algorithms
+ */
 int fib(int n) {
+    if (n < 0) {
+        return -1;
+    }
+    if (n == 0) {
+        return 0;
+    }
     int res[3];
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:34]: Let's use a C++-style array (`std::vector`) instead. This will also fix the CI warnings.
The `vector` library/header must be added in order for this to work.
```suggestion
    std::vector<int> res(3);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: dynamic_programming/fibonacci_bottom_up.cpp [hunk 11]
```diff
@@ -1,7 +1,41 @@
...
+ */
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for storing the values of nth fibonacci number
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
 int fib(int n) {
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_bottom_up.cpp:31]: This function should be documented explaining the use of its parameters, a brief description, and what it returns.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
